src/pybind/mgr/dashboard/frontend/src/app/shared/directives/auth-storage.directive.ts

```diff
@@ -0,0 +1,93 @@
+import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
+
+import _ from 'lodash';
+
+import { Permission, Permissions } from '~/app/shared/models/permissions';
+import { AuthStorageService } from '~/app/shared/services/auth-storage.service';
+
+@Directive({
+  selector: '[cdAuthStorage]'
+})
+export class AuthStorageDirective {
+  permissions: Permissions;
+
+  constructor(
+    private templateRef: TemplateRef<any>,
+    private viewContainer: ViewContainerRef,
+    private authStorageService: AuthStorageService
+  ) {}
+
+  @Input('cdAuthStorage') set cdAuthStorage(condition: string | string[] | object) {
+    this.permissions = this.authStorageService.getPermissions();
+    if (this.isAuthorized(condition)) {
+      this.viewContainer.createEmbeddedView(this.templateRef);
+    } else {
+      this.viewContainer.clear();
+    }
+  }
+
+  private isAuthorized(condition: any): boolean {
+    /*
+      single permission equivalent to check e.g osd.
+      We check if the permission key exists on the Permissions class then
+      if true We check for read permission and display the component accordingly
+    */
+
+    if (_.isString(condition)) {
+      if (_.has(this.permissions, [condition, 'read']) && this.permissions[condition]['read']) {
+        return true;
+      }
+    }
+
+    /*
+      lodash _isArray matches both objects and arrays.
+      By checking if the first index contains a value we can tell if it is truly an array or not
+      Also works for _isObject
+    */
+
+    if (_.isArray(condition) && condition[0] !== undefined) {
+      const checkedPermissions: boolean[] = [];
+
+      _.forEach(condition, (value) => {
+        if (_.has(this.permissions, [value, 'read']) && this.permissions[value]['read']) {
+          checkedPermissions.push(true);
+        } else {
+          checkedPermissions.push(false);
+        }
+      });
+
+      /*
+        check if all permissions in the checkedPermissions array are true.
+        if no false values are found then requested permissions are granted so we display the component
+      */
+      if (!checkedPermissions.includes(false)) {
+        return true;
+      }
+    }
```

Comment:
Here you can replace all this code by using lodash [`some()`](https://lodash.com/docs/4.17.15#some):
```ts
return _.some(this.permissions, ...)
```

Reply:
Finally understood how to use _some!. I found it was not the perfect fit in this case because it stops evaluating once it meets the first truthy value. Arrays with falsy permissions still passed.

I found  _every more suitable instead. It evaluates all permissions and only returns true if all are true.

Reply:
Here you have to decide which case is the one you have to implement:
* Strict: ALL conditions have to be met (`_.every()` in JS, `all()` in Python) 
* Loose: At least ONE condition has to be met (`_.some()` in JS, `any()` in Python).

In this very case, if a developer tags part of the code with specific permissions:
```ts
<cdNewComponent *cdScope="['osd', 'pool']">
...
</cdNewComponent>
```
What is she/he meaning here? That the user is expected to have all permissions at the same time, or is it enough with any of them?

I'd say it's the second, because if the developer wants to restrict a specific part of the code for a give permission, they would rather write something like:
```ts
<cdNewComponent *cdScope="['osd', 'pool']">
  <cdNewOSDSubComponent *cdScope="'osd'">
    This Subcomponent can only be accessed by users with OSD role/permission 
  </cdNewOSDSubComponent>
  <cdNewPoolSubComponent *cdScope="'pool'">
    This Subcomponent can only be accessed by users with Pool role/permission
  </cdNewPoolSubComponent> 
...
</cdNewComponent>
```

Reply:
@epuertat in fact we need both modes (The OR and the AND) for a full replacement (currently we are only replacing in:
src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-list/pool-list.component.html).
The code is assuming the AND mode as the default.
As we need a follow-up PR replacing the permissions check with the directive, we can decide which should be the default later.

Reply:
Hello @epuertat, @alfonsomthd not sure what I am to do here

Reply:
Hello @epuertat, @alfonsomthd not sure what I am to do here

Reply:
You're right @alfonsomthd !

@nSedrickm so in this very case we need [a structural directive with multiple inputs](https://stackoverflow.com/questions/41789702/how-to-use-angular-structural-directive-with-multiple-inputs) (this is boss level Angular: please mind the names of the input variables).

I'd suggest something like `<cdComponent *cdScope="['osd', 'pool']; matchAll:true">`. By default if any permission matches, the directive displays the component. If `matchAll:true` is specified, then all permissions need to match.

What do you think @nSedrickm ?

Reply:
(me scratching head) 

@epuertat I see this will enable the developer to specify whether to match all or some permissions. Nice I'm guessing matchAll will only need to be specified for array and object permissions with its default case being false.

I'll work on it

Reply:
Yeah, that's exactly it! Thanks a lot @nSedrickm !